src/workerd/jsg/memory.h

```diff
@@ -0,0 +1,810 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+// Implements mechanism for incorporating details about the native (c++) objects
+// in a v8 heap snapshot. The design of the API and implementation were heavily
+// influenced by Node.js' implementation of the same feature.
+
```

Comment:
This header is pretty large, can/should a lot of the code in here be moved into a .c++ file?

I wouldn't let that block the merging of this PR though.

Reply:
Most of this are inline template method definitions. Will explore ways of simplifying in another iteration.